Index: src/getdata.h.in
===================================================================
--- src/getdata.h.in	(Revision 448)
+++ src/getdata.h.in	(Arbeitskopie)
@@ -207,6 +207,15 @@
 #undef _gd_int16_t
 #undef _gd_uint16_t
 #undef _gd_int64_t
+
+#if defined(__MINGW32__) || defined(_MSC_VER)
+// TODO move to config.h file
+typedef short _gd_int16_t;
+typedef unsigned short _gd_uint16_t;
+typedef __int64 _gd_int64_t;
+#endif
+
+
 #if ! defined GD_C89_API && ! defined GD_NO_C99_API
 #  include <inttypes.h>
 #  define GD_DCOMPLEXM(v) double _Complex v
